### ISSUE TITLE
Implement document generation tool like Sphinx / Doxygen #12

### ISSUE THREAD

**Issue by benfitzpatrick** [metoppv/improver#13]
Do after #12.

As an IMPROVER developer, I want a document generation tool to be chosen and implemented so that I can write my docstrings accordingly and create automatic code documentation.

Acceptance criteria:
 * Implement choice of document generation tool from #12.
 * Write automated test so that document generation passes or fails.

**Comment by arh89** [comment]
For first deployment of Sphinx documentation, see: http://improver.readthedocs.io/

This builds automatically each time there is a commit to my "Sphinx_test" branch: https://github.com/arh89/improver/tree/Sphinx_test
(Although once the team has agreed, this will be merged into metoppv/improver/master, and a new commit to that branch will trigger a rebuild).

Following a similar method to Travis CI, there are now .readthedocs.yml and environment.yml files which define and use a conda environment which allows readthedocs to build iris (and its dependencies) before building our documentation. This is required because of the way Sphinx parses the documentation - it does so by actually importing our code (in python), rather than just statically reading the source files.

## To do:
* Set up some documentation testing - so that we know when the documentation fails to build. (I know that we can generate a badge which can go into the readme.md, but maybe we want more than this? - @benfitzpatrick?)
* Adjust formatting of documentation as necessary (following feedback from dev team)
* Look into intersphinx links : we can automatically link to relevant parts of the iris (and numpy etc) documentation from within ours (a big plus for using Sphinx).

**Comment by benfitzpatrick** [comment]
The documentation looks good. I think the only thing that would be good is that I would like something for Travis to install sphinx and attempt to build the documentation in a 'BATS' test

**Comment by arh89** [comment]
I will look into how to do this. Since Travis needs to install iris anyway, I don't think there should be any significant issues (because most of the dependencies should have already been met), nor should there be any significant additional overhead (to Travis: thinking about speed of build/test).

That said, I do wonder if there might be a simpler route than employing a second build of the documentation...

Currently, we are able to query readthedocs to see if the documentation build passed (although this can be slow). As a slightly more lightweight, but hopefully still useful test, perhaps we can employ some sort of test which simply checks that the docstrings are aligned to some standard (numpy/google style)? What are your thoughts on this? Is there an advantage to doing this check with Travis rather than relying on buildthedocs?

**Comment by arh89** [comment]
## Looking into automatically generating *.rst files:
It seems relatively straightforward to get Travis to run `sphinx-apidoc` (or `cd doc && make apidoc` to automatically generate the API documentation) as part of the test suite, but I am not sure of a _straightforward_ way to get readthedocs to run this command before it builds the documentation.

There are a few options (none of which are ideal, in my opinion):
    1. Continue along the path of requiring a `make apidoc` whenever a new module/file is added (as part of the definition of done?), and store the *.rst files in the repository. i.e: Don't get `sphinx-apidoc` to run automatically.
    2. Create a root setup.py file, which checks the `READTHEDOCS` environment variable, and runs this (+ possibly installing packages if this means we can't use the built-in RTD conda stuff).
    3. Add this to conf.py. I noticed that other people have been doing this successfully: https://github.com/rtfd/readthedocs.org/issues/1139 but I don't like this idea, because I don't know internally how sphinx-apidoc/sphinx in general will parse/run conf.py - I worry that this might break in the future.
    4. Travis can be given force push access (see: https://github.com/metomi/rose/pull/2046#issuecomment-288132630) which can be used to run this. Since RTD will only update once master has been updated, this can be used as a workaround. In my opinion this can be quite dangerous!

Another possibility, would be to stick with 1), and mandate that developers run `sphinx-apidoc` to generate their documentation, and then use Travis as a means of checking (by running this command) that none of the files have been updated/added. I think this might possibly be more effort than it's worth though, especially to get it right (in the case of: excluded files and/or having some files which have been edited by the developer).

In my opinion, since `sphinx-apidoc` does not generate perfect documentation anyway, and we will likely want to use this as a base and manually edit the files at some point in the future (to improve the quality of the documentation), I would suggest going with option 1).

**Comment by benfitzpatrick** [comment]
I have to say, I like option 3) - if it breaks later, we can always fix it.

**Comment by arh89** [comment]
Preliminary test suggests that option 3) seems to work just fine. After looking at the sphinx source code, it seems that many of the extensions work in a similar method (in effect, adding this to conf.py is a bit like defining our own extension) - this gives me some confidence that things hopefully won't break in the future.

**Comment by arh89** [comment]
**PR:**
https://github.com/metoppv/improver/pull/88

Initially this PR was created just to test that Travis could build the documentation (it was easier than setting up Travis for my own repository), but actually, since everything appears to work, I'd say that this is now ready for review.

**Comment by arh89** [comment]
I just noticed that the RTD hosted documentation has an "Edit on Github" link which results in a 404 error on most pages - this is because they link to the *.rst files in the repo, which are not present because they are instead generated at build time.

Note: Local builds use the Sphinx default theme (rather than the RTD theme), and so this won't be a problem.

Apparently, we can fix this problem by copying one of the template files from the RTD theme to our repo:
https://stackoverflow.com/questions/36019670/removing-the-edit-on-github-link-when-using-read-the-docs-sphinx-with-readthed

...I think that this fix is quite messy though. Including a single file to fix something which is not actually a part of the repo doesn't seem like an elegant solution.

**Comment by benfitzpatrick** [comment]
@arh89, is this closed now?

**Comment by arh89** [comment]
@benfitzpatrick, since everything seems to be working OK, I'd say it's safe to close this.
